File: js/parser.js
/**
 * Marila Language Parser
 * Transforms tokens into an AST for video, document, or image file blocks.
 */

class Parser {
  constructor(tokens) {
    this.tokens = tokens;
    this.pos = 0;
  }

  peek() {
    return this.tokens[this.pos];
  }

  advance() {
    const token = this.peek();
    this.pos++;
    return token;
  }

  match(type, val = null) {
    const t = this.peek();
    if (!t) return false;
    if (t.type === type && (val === null || t.value.toLowerCase() === val.toLowerCase())) {
      this.advance();
      return true;
    }
    return false;
  }

  expect(type, val = null) {
    const t = this.peek();
    if (!t) throw new Error(`Unexpected end of input, expected ${val || type}`);
    if (t.type === type && (val === null || t.value.toLowerCase() === val.toLowerCase())) {
      return this.advance();
    }
    throw new Error(`Line ${t.line}, Col ${t.col}: Expected ${val || type} but got '${t.value}'`);
  }

  parse() {
    // Top-level structure:
    // file type <type>: [ ... ] render <type>
    let fileType = 'video';

    if (this.match(TokenType.KEYWORD, 'file')) {
      this.expect(TokenType.KEYWORD, 'type');
      const typeTok = this.advance();
      fileType = typeTok.value.toLowerCase();
      if (this.peek() && this.peek().value === ':') {
        this.advance();
      }
    }

    // Expect opening bracket '['
    if (this.peek() && this.peek().value === '[') {
      this.advance();
    }

    const statements = [];
    while (this.pos < this.tokens.length) {
      const t = this.peek();
      if (!t || t.type === TokenType.EOF) break;
      if (t.value === ']') {
        this.advance();
        break;
      }

      const stmt = this.parseStatement(fileType);
      if (stmt) {
        statements.push(stmt);
      } else {
        // Skip unknown token to prevent infinite loop
        this.advance();
      }
    }

    // Optional: render <type>
    if (this.match(TokenType.KEYWORD, 'render')) {
      if (this.peek() && (this.peek().type === TokenType.KEYWORD || this.peek().type === TokenType.IDENTIFIER)) {
        this.advance();
      }
    }

    return {
      type: 'Program',
      fileType,
      statements
    };
  }

  parseStatement(fileType) {
    if (fileType === 'document') {
      return this.parseDocumentStatement();
    } else if (fileType === 'video') {
      return this.parseVideoStatement();
    } else if (fileType === 'image') {
      return this.parseImageStatement();
    }
    return this.parseGenericStatement();
  }

  // Parse Document Statement (e.g. p1 - "Header" :: Subtitle or p2 - "Body text <math>")
  parseDocumentStatement() {
    const t = this.peek();
    if (!t) return null;

    let id = null;
    if (t.type === TokenType.IDENTIFIER || t.type === TokenType.KEYWORD) {
      id = this.advance().value;
      if (this.peek() && this.peek().value === '-') {
        this.advance(); // consume '-'
      }
    }

    // Content string or text
    let text = '';
    if (this.peek() && this.peek().type === TokenType.STRING) {
      text = this.advance().value;
    } else {
      // Gather unquoted text until newline or symbol
      while (this.peek() && this.peek().value !== '::' && this.peek().value !== ']' && this.peek().type !== TokenType.EOF) {
        text += (text ? ' ' : '') + this.advance().value;
      }
    }

    // Optional subtitle after '::'
    let subtitle = null;
    if (this.peek() && this.peek().value === '::') {
      this.advance(); // consume ::
      if (this.peek() && this.peek().type === TokenType.STRING) {
        subtitle = this.advance().value;
      } else {
        let subText = '';
        while (this.peek() && this.peek().value !== ']' && this.peek().type !== TokenType.EOF) {
          subText += (subText ? ' ' : '') + this.advance().value;
        }
        subtitle = subText;
      }
    }

    return {
      kind: 'DocumentNode',
      id: id || 'p',
      text: text,
      subtitle: subtitle
    };
  }

  // Parse Video Statement (e.g. square(size 30) morphs to circle(size 10) or plot sin(x))
  parseVideoStatement() {
    const t = this.peek();
    if (!t) return null;

    // Check if statement contains "morphs to"
    const shape1 = this.parseShapeCall();
    if (shape1) {
      if (this.match(TokenType.KEYWORD, 'morphs')) {
        this.expect(TokenType.KEYWORD, 'to');
        const shape2 = this.parseShapeCall();
        let duration = '3s';
        if (this.match(TokenType.KEYWORD, 'over')) {
          duration = this.advance().value;
        }
        return {
          kind: 'MorphAnimation',
          fromShape: shape1,
          toShape: shape2,
          duration: duration
        };
      }
      return {
        kind: 'StaticShape',
        shape: shape1
      };
    }

    // Check plot or function statement
    if (this.match(TokenType.KEYWORD, 'plot') || this.match(TokenType.KEYWORD, 'draw')) {
      let funcExpr = '';
      while (this.peek() && this.peek().value !== ']' && this.peek().value !== '&' && this.peek().type !== TokenType.EOF) {
        funcExpr += (funcExpr ? ' ' : '') + this.advance().value;
      }
      let animateTrace = false;
      if (this.peek() && this.peek().value === '&') {
        this.advance();
        while (this.peek() && this.peek().value !== ']' && this.peek().type !== TokenType.EOF) {
          if (this.peek().value.toLowerCase() === 'animate' || this.peek().value.toLowerCase() === 'trace') {
            animateTrace = true;
          }
          this.advance();
        }
      }
      return {
        kind: 'PlotAnimation',
        expression: funcExpr || 'x^2',
        animateTrace: animateTrace
      };
    }

    // Fallback unparsed video command
    let lineText = '';
    while (this.peek() && this.peek().value !== ']' && this.peek().type !== TokenType.EOF) {
      lineText += (lineText ? ' ' : '') + this.advance().value;
    }
    return { kind: 'RawCommand', text: lineText };
  }

  // Helper: parse shape call like square(size 30, color "cyan") or point A at (0, 0)
  parseShapeCall() {
    const t = this.peek();
    if (!t) return null;
    const name = t.value.toLowerCase();
    const validShapes = ['square', 'circle', 'triangle', 'rectangle', 'point', 'line', 'polygon', 'axes', 'star'];
    
    if (t.type === TokenType.KEYWORD || t.type === TokenType.IDENTIFIER) {
      if (validShapes.includes(name)) {
        this.advance();
        const args = {};
        if (this.peek() && this.peek().value === '(') {
          this.advance(); // consume '('
          while (this.peek() && this.peek().value !== ')') {
            const key = this.advance().value;
            let val = true;
            if (this.peek() && (this.peek().type === TokenType.NUMBER || this.peek().type === TokenType.STRING || this.peek().type === TokenType.IDENTIFIER)) {
              val = this.advance().value;
            }
            args[key] = val;
            if (this.peek() && this.peek().value === ',') this.advance();
          }
          if (this.peek() && this.peek().value === ')') this.advance();
        }
        return { name, args };
      }
    }
    return null;
  }

  // Parse Image (TikZ) Statement (e.g. point A at (0, 0), line from A to B, circle at C radius 2)
  parseImageStatement() {
    const t = this.peek();
    if (!t) return null;

    if (this.match(TokenType.KEYWORD, 'point')) {
      const id = this.expect(TokenType.IDENTIFIER).value;
      let x = 0, y = 0;
      if (this.match(TokenType.KEYWORD, 'at') || (this.peek() && this.peek().value === '(')) {
        if (this.peek() && this.peek().value === '(') this.advance();
        x = this.expect(TokenType.NUMBER).value;
        if (this.peek() && this.peek().value === ',') this.advance();
        y = this.expect(TokenType.NUMBER).value;
        if (this.peek() && this.peek().value === ')') this.advance();
      }
      let label = id;
      if (this.match(TokenType.KEYWORD, 'label')) {
        label = this.advance().value;
      }
      return { kind: 'Point', id, x, y, label };
    }

    if (this.match(TokenType.KEYWORD, 'line') || this.match(TokenType.KEYWORD, 'segment')) {
      let from = 'A', to = 'B';
      if (this.match(TokenType.KEYWORD, 'from')) from = this.advance().value;
      if (this.match(TokenType.KEYWORD, 'to')) to = this.advance().value;
      return { kind: 'Line', from, to };
    }

    if (this.match(TokenType.KEYWORD, 'triangle')) {
      let pts = [];
      const idStr = this.advance().value; // e.g. ABC
      pts = idStr.split('');
      let color = 'amber';
      let showAngles = false;
      
      // Parse parameters after triangle
      while (this.peek() && this.peek().value !== ']' && this.peek().type !== TokenType.EOF) {
        const val = this.peek().value.toLowerCase();
        if (val === 'color') {
          this.advance();
          color = this.advance().value;
        } else if (val === 'show' || val === 'angles') {
          showAngles = true;
          this.advance();
        } else {
          this.advance();
        }
      }
      return { kind: 'Triangle', points: pts, color, showAngles };
    }

    if (this.match(TokenType.KEYWORD, 'circle')) {
      let center = 'A';
      let radius = 50;
      let dash = false;

      while (this.peek() && this.peek().value !== ']' && this.peek().type !== TokenType.EOF) {
        const val = this.peek().value.toLowerCase();
        if (val === 'at') {
          this.advance();
          center = this.advance().value;
        } else if (val === 'radius') {
          this.advance();
          radius = this.expect(TokenType.NUMBER).value;
        } else if (val === 'dash' || val === 'dashed') {
          dash = true;
          this.advance();
        } else {
          this.advance();
        }
      }
      return { kind: 'Circle', center, radius, dash };
    }

    // Generic fallback image node
    let lineText = '';
    while (this.peek() && this.peek().value !== ']' && this.peek().type !== TokenType.EOF) {
      lineText += (lineText ? ' ' : '') + this.advance().value;
    }
    return { kind: 'GenericDiagram', text: lineText };
  }

  parseGenericStatement() {
    const t = this.advance();
    return { kind: 'Generic', value: t ? t.value : '' };
  }
}

if (typeof module !== 'undefined') {
  module.exports = { Parser };
}

File: marila/parser.py
"""
Marila Programming Language - Recursive Descent Parser
"""

from marila.tokens import Token, TokenType
from marila.ast_nodes import (
    ProgramNode, DocumentNode, MorphNode, PlotNode,
    PointNode, LineNode, TriangleNode, CircleNode
)

class ParserError(Exception):
    def __init__(self, message: str, token: Token = None):
        if token:
            super().__init__(f"ParserError at Line {token.line}, Col {token.col}: {message} (got '{token.value}')")
        else:
            super().__init__(f"ParserError: {message}")

class Parser:
    def __init__(self, tokens: list[Token]):
        self.tokens = tokens
        self.pos = 0

    def peek(self, offset: int = 0) -> Token:
        idx = self.pos + offset
        return self.tokens[idx] if idx < len(self.tokens) else self.tokens[-1]

    def advance() -> Token:
        tok = self.peek()
        self.pos += 1
        return tok

    def match(type_: TokenType, value: str = None) -> bool:
        tok = self.peek()
        if tok.type == type_ and (value is None or tok.value.lower() == value.lower()):
            self.advance()
            return True
        return False

    def expect(type_: TokenType, value: str = None) -> Token:
        tok = self.peek()
        if tok.type == type_ and (value is None or tok.value.lower() == value.lower()):
            return self.advance()
        raise ParserError(f"Expected {value or type_.name}", tok)

    def parse() -> ProgramNode:
        file_type = "video"

        # file type <type>: [ ... ] render <type>
        if self.match(TokenType.KEYWORD, "file"):
            self.expect(TokenType.KEYWORD, "type")
            file_type_tok = self.advance()
            file_type = file_type_tok.value.lower()
            if self.peek().type == TokenType.COLON:
                self.advance()

        if self.peek().type == TokenType.LBRACKET:
            self.advance()

        statements = []
        while self.pos < len(self.tokens):
            tok = self.peek()
            if tok.type == TokenType.EOF or tok.type == TokenType.RBRACKET:
                if tok.type == TokenType.RBRACKET:
                    self.advance()
                break

            stmt = self.parse_statement(file_type)
            if stmt:
                statements.append(stmt)
            else:
                self.advance()

        render_target = None
        if self.match(TokenType.KEYWORD, "render"):
            render_target = self.advance().value

        return ProgramNode(file_type=file_type, statements=statements, render_target=render_target)

    def parse_statement(file_type: str):
        if file_type == "document":
            return self.parse_document_statement()
        elif file_type == "video":
            return self.parse_video_statement()
        elif file_type == "image":
            return self.parse_image_statement()
        return None

    def parse_document_statement() -> DocumentNode:
        tok = self.peek()
        node_id = "p"

        if tok.type in (TokenType.IDENTIFIER, TokenType.KEYWORD):
            node_id = self.advance().value
            if self.peek().type == TokenType.MINUS:
                self.advance()

        text = ""
        if self.peek().type == TokenType.STRING:
            text = self.advance().value
        else:
            words = []
            while self.peek().type not in (TokenType.DOUBLE_COLON, TokenType.RBRACKET, TokenType.EOF):
                words.append(self.advance().value)
            text = " ".join(words)

        subtitle = None
        if self.peek().type == TokenType.DOUBLE_COLON:
            self.advance() # consume ::
            if self.peek().type == TokenType.STRING:
                subtitle = self.advance().value
            else:
                sub_words = []
                while self.peek().type not in (TokenType.RBRACKET, TokenType.EOF):
                    sub_words.append(self.advance().value)
                subtitle = " ".join(sub_words)

        return DocumentNode(node_id=node_id, text=text, subtitle=subtitle)

    def parse_video_statement():
        # Check for morph animation: shape(args) morphs to shape(args)
        shape1 = self.parse_shape_call()
        if shape1:
            if self.match(TokenType.KEYWORD, "morphs"):
                self.expect(TokenType.KEYWORD, "to")
                shape2 = self.parse_shape_call()
                duration = "3s"
                if self.match(TokenType.KEYWORD, "over"):
                    duration = self.advance().value
                return MorphNode(from_shape=shape1, to_shape=shape2, duration=duration)

        if self.match(TokenType.KEYWORD, "plot") or self.match(TokenType.KEYWORD, "draw"):
            expr_words = []
            while self.peek().type not in (TokenType.AMPERSAND, TokenType.RBRACKET, TokenType.EOF):
                expr_words.append(self.advance().value)
            expr = " ".join(expr_words) or "sin(x)"
            
            animate_trace = True
            if self.peek().type == TokenType.AMPERSAND:
                self.advance()
                while self.peek().type not in (TokenType.RBRACKET, TokenType.EOF):
                    self.advance()

            return PlotNode(expression=expr, animate_trace=animate_trace)

        return None

    def parse_shape_call() -> dict:
        tok = self.peek()
        if tok.type in (TokenType.KEYWORD, TokenType.IDENTIFIER):
            name = tok.value.lower()
            if name in ('square', 'circle', 'triangle', 'rectangle', 'axes'):
                self.advance()
                args = {}
                if self.peek().type == TokenType.LPAREN:
                    self.advance()
                    while self.peek().type != TokenType.RPAREN and self.peek().type != TokenType.EOF:
                        key = self.advance().value
                        val = True
                        if self.peek().type in (TokenType.NUMBER, TokenType.STRING, TokenType.IDENTIFIER):
                            val = self.advance().value
                        args[key] = val
                        if self.peek().type == TokenType.COMMA:
                            self.advance()
                    if self.peek().type == TokenType.RPAREN:
                        self.advance()
                return {'name': name, 'args': args}
        return None

    def parse_image_statement():
        if self.match(TokenType.KEYWORD, "point"):
            point_id = self.expect(TokenType.IDENTIFIER).value
            x, y = 0.0, 0.0
            if self.match(TokenType.KEYWORD, "at") or self.peek().type == TokenType.LPAREN:
                if self.peek().type == TokenType.LPAREN: self.advance()
                x = float(self.expect(TokenType.NUMBER).value)
                if self.peek().type == TokenType.COMMA: self.advance()
                y = float(self.expect(TokenType.NUMBER).value)
                if self.peek().type == TokenType.RPAREN: self.advance()

            label = point_id
            if self.match(TokenType.KEYWORD, "label"):
                label = self.advance().value
            return PointNode(point_id=point_id, x=x, y=y, label=label)

        if self.match(TokenType.KEYWORD, "line") or self.match(TokenType.KEYWORD, "segment"):
            from_id = "A"
            to_id = "B"
            if self.match(TokenType.KEYWORD, "from"): from_id = self.advance().value
            if self.match(TokenType.KEYWORD, "to"): to_id = self.advance().value
            return LineNode(from_id=from_id, to_id=to_id)

        if self.match(TokenType.KEYWORD, "triangle"):
            pts = list(self.advance().value)
            return TriangleNode(point_ids=pts, color="amber", show_angles=True)

        if self.match(TokenType.KEYWORD, "circle"):
            center_id = "A"
            radius = 50.0
            dash = False
            while self.peek().type not in (TokenType.RBRACKET, TokenType.EOF):
                val = self.peek().value.lower()
                if val == "at":
                    self.advance()
                    center_id = self.advance().value
                elif val == "radius":
                    self.advance()
                    radius = float(self.expect(TokenType.NUMBER).value)
                elif val in ("dash", "dashed"):
                    dash = True
                    self.advance()
                else:
                    self.advance()
            return CircleNode(center_id=center_id, radius=radius, dash=dash)

        return None
